addons/hr_attendance/models/hr_attendance.py

```diff
@@ -92,71 +92,45 @@ def _compute_color(self):
-                                att_progress_values[attendance[0]] = 0
-                                overtime_reservoir -= attendance[1]
-                            else:
-                                att_progress_values[attendance[0]] = float(((attendance[1] - overtime_reservoir) / attendance[1]) * 100)
-                                overtime_reservoir = 0
-                        else:
-                            att_progress_values[attendance[0]] = 100
-                elif overtime_reservoir < 0 and grouped_dict[overtime]['attendances']:
-                    att_id = grouped_dict[overtime]['attendances'][0][0]
-                    att_progress_values[att_id] = overtime_reservoir
-                    negative_overtime_attendances[att_id] = True
-        for attendance in self:
-            if negative_overtime_attendances[attendance.id]:
-                attendance.overtime_hours = att_progress_values.get(attendance.id, 0)
-            else:
-                attendance.overtime_hours = attendance.worked_hours * ((100 - att_progress_values.get(attendance.id, 100)) / 100)
+            day_starts = [self._get_day_start_and_day(att.employee_id, att.check_in) for att in self]
+            date_min, date_max = min(date for dummy, date in day_starts), max(date for dummy, date in day_starts)
+            time_min, time_max = min(time for time, dummy in day_starts), max(time for time, dummy in day_starts)
+
+            # Find at least all attendances that happen on the same day and for the same employee
+            # as an attendance in self
+            all_attendances = self | self.env['hr.attendance'].search(
+                [
+                    ('employee_id', 'in', self.employee_id.ids),
+                    ('check_in', '>=', time_min),
+                    ('check_in', '<=', time_max + timedelta(days=1)),
+                ],
+            )
+            attendances_by_date_and_employee = defaultdict(list)
+            for att in all_attendances:
+                dummy, date = att._get_day_start_and_day(att.employee_id, att.check_in)
+                # reset affected attendances
+                if date_min <= date <= date_max:
+                    att.overtime_hours = 0
+                    attendances_by_date_and_employee[date, att.employee_id.id].append(att)
+
+            overtimes = self.env['hr.attendance.overtime'].search(
+                [
+                    ('employee_id', 'in', self.employee_id.ids),
+                    ('date', '>=', date_min),
+                    ('date', '<=', date_max),
+                ],
+            )
+            for ot in overtimes:
+                # Distribute overtime to attendances
+                overtime_reserve = ot.duration
+                attendances = attendances_by_date_and_employee[ot.date, ot.employee_id.id]
+                for att in sorted(attendances, key=lambda att: att.check_in, reverse=True):
+                    if overtime_reserve == 0:
+                        break
```

Comment:
iirc the problem was that there isn't always an overtime -- for example if we're reducing an attendance and the overtime gets deleted -- and since this loop is on overtimes it's cleaner (imho) to do it in a separate step here: https://github.com/odoo/odoo/blob/23f3d8fbbcf85d868c49222bbbe93e039dd56c34/addons/hr_attendance/models/hr_attendance.py#L112-L115